drivers/can/can_mcan.c

```diff
@@ -1122,9 +1112,9 @@ int can_mcan_add_rx_filter(const struct device *dev, can_rx_callback_t callback,
 
 #ifdef CONFIG_CAN_FD_MODE
 	if ((filter->flags &
-	     ~(CAN_FILTER_IDE | CAN_FILTER_DATA | CAN_FILTER_RTR | CAN_FILTER_FDF)) != 0U) {
+	     ~(CAN_FILTER_IDE | CAN_FILTER_DATA | CAN_FILTER_FDF)) != 0U) {
 #else  /* CONFIG_CAN_FD_MODE */
-	if ((filter->flags & ~(CAN_FILTER_IDE | CAN_FILTER_DATA | CAN_FILTER_RTR)) != 0U) {
+	if ((filter->flags & ~(CAN_FILTER_IDE | CAN_FILTER_DATA)) != 0U) {
```

Comment:
The `DATA` flag is the inverse of the `RTR` flag. Should we not either allow setting both flags or none?

Reply:
Setting none of them would make no sense. What types of frames would that match?

Reply:
Having to set both flags just to get data frames would also be kind of odd and create another set of problems.

Reply:
Before the flags were introduced in f8a88cdb2791fe40b3021f1275fe250c2c17dcf3 we had an `rtr_mask` bit in the filter struct. If this bit was set to `0`, both data and RTR frames would be matched. In order to achieve the same behavior with flags, we need to set the flags `CAN_FILTER_DATA | CAN_FILTER_RTR`. In my understanding, if I configure a filter with `flags` set to `CAN_FILTER_DATA` only, it should never match an RTR frame. However, exactly this would happen with this change, if I'm not wrong.

My expectation for ` can_add_rx_filter` would be to return `-ENOTSUP` if I set the `CAN_FILTER_DATA` flag only and the driver is not able to return only data frames.

My initial idea was to allow setting none of both flags to indicate "don't care" as well, but this would indeed not make much sense.